### PR TITLE
Set Payment Date for auto created Paid Payment

### LINKED ISSUE #2869: Auto created Paid Payment when Opp updated from null to value is populating Scheduled Date instead of Payment Date
NPSP 3.115

STR: Create Closed/Won opp, leave amount blank. Update Opportunity to a value. Note Payment created is properly checked as 'Paid' however Scheduled Date is populated rather than Payment Date. Scheduled Date should be blank, Payment Date should be Today.

image

### PR DESCRIPTION
**Pull request by kofijohnson**
# Critical Changes

# Changes

# Issues Closed
Fixes #2869

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: src/classes/PMT_PaymentCreator.cls [hunk 1]
```diff
@@ -349,9 +349,15 @@ public class PMT_PaymentCreator {
...
             npe01__Paid__c = newOpportunity.IsClosed,
-            npe01__Scheduled_Date__c = newOpportunity.CloseDate,
             npe01__Payment_Amount__c = newOpportunity.Amount
-        );                                
+        );
+
+        if (opportunityPayment.npe01__Paid__c) {
+            opportunityPayment.npe01__Payment_Date__c = newOpportunity.CloseDate;
+        }
+        else {
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator.cls]: Update the formatting of the else statement to -- } else {

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 2]
```diff
@@ -176,205 +176,270 @@ private with sharing class PMT_PaymentCreator_TEST {
...
-        System.assertEquals(1, createdPayments.size(), 'new payment should be created.');
-        System.assertEquals(testOpportunity.CloseDate, createdPayments[0].npe01__Scheduled_Date__c);
-        System.assertEquals(testOpportunity.Amount, createdPayments[0].npe01__Payment_Amount__c);
-        System.assertEquals(true, createdPayments[0].npe01__Paid__c);
+        createdPayments = getCreatedPayments(createdTestOpportunities.keySet());
+        
+        System.assertEquals(testOpportunitiesCount, createdPayments.size(), testOpportunitiesCount + ' new payments should be created.');
+        for (npe01__OppPayment__c createdPayment : createdPayments) {
+            Opportunity paymentOpportunity = createdTestOpportunities.get(createdPayment.npe01__Opportunity__c);
+            System.assert(paymentOpportunity != null);
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator_TEST.cls]: What are your thoughts about using assertNotEquals in place of assert?

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 3]
```diff
@@ -176,205 +176,270 @@ private with sharing class PMT_PaymentCreator_TEST {
...
-        ];
+        createdPayments = getCreatedPayments(createdTestOpportunities.keySet());
         System.assertEquals(0, createdPayments.size(), 'No payment should be created because the opportunity is Closed Lost');
     }
 
     /*********************************************************************************************************
     * @description
     operation:
-        Update a Closed Lost Opportunity Amount from zero to a value
+        Update a List of Closed Lost Opportunities Amount from zero to a value
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator_TEST.cls]: Lowercase the "L" of List.

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 4]
```diff
@@ -176,205 +176,270 @@ private with sharing class PMT_PaymentCreator_TEST {
...
+            System.assertEquals(paymentOpportunity.Amount, createdPayment.npe01__Payment_Amount__c);
+            System.assertEquals(false, createdPayment.npe01__Paid__c);
+        } 
     }
 
     /*********************************************************************************************************
     * @description
     operation:
-        Update a Closed Lost Opportunity Amount from null to a value
+        Update a List of Closed Lost Opportunities Amount from null to a value
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator_TEST.cls]: Lowercase the "L" of List.

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 5]
```diff
@@ -176,205 +176,270 @@ private with sharing class PMT_PaymentCreator_TEST {
...
     verify:
-        No payment should be created because the opportunity is closed lost.
-    **********************************************************************************************************/            
+        No payment should be created because the opportunities are closed lost.
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator_TEST.cls]: Uppercase "o" of opportunities" and the "c" & "l" of closed lost.

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 6]
```diff
@@ -176,205 +176,270 @@ private with sharing class PMT_PaymentCreator_TEST {
...
+        }
+        insert testOpportunities;
 
-        return testOpportunity;
+        return new Map<Id, Opportunity>(testOpportunities);
+    }
+
+    /*********************************************************************************************************
+    * @description Gets the list of created payments.
+    * @param testOpportunitiesIds The Id of test opportunities.
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator_TEST.cls]: Update Id to Ids

**Comment by kofijohnson** [comment]
Thank you @joy-sfdo. It is fixed now.

**Pull request merged**